### DISCUSSION TITLE
Change the unit of delay in stages from milliseconds to seconds

### DISCUSSION THREAD

**Discussion by yuanchen8911** [kubernetes-sigs/kwok#1070] [Ideas]
The delay field in a stage file is specified in milliseconds. I suggest changing it to seconds.
- Do we really need this level of granularity?
- We want to set a very large delay, but the default maximum value in YAML is 999,999. 999,999 milliseconds is too short for our tests.
/cc @dmitsh

**Comment by dmitsh** [comment]
The problem I'm trying to solve is not related to the delay.
I'm trying to start a pod and keep it running until I decide to terminate it.
With the current stages, the pod becomes "Completed" right after "Running".

**Reply by wzshiming** [reply]
Looks like a Pod created by a Job? Maybe you should delete the Stage or add another match rule to the Stage?

https://github.com/kubernetes-sigs/kwok/blob/main/kustomize/stage/pod/fast/pod-complete.yaml

**Reply by yuanchen8911** [reply]
The delay parameters in the pod-complete stage determine how fast the pod will go from  `Running` to `Completed`, aren't they?  That's why I proposed the introduction of configurable delays.

**Reply by dmitsh** [reply]
Sometimes it is not possible to estimate how long the pod will run, especially for complex scenarios such as preemption.
We solved this issue by removing Job kind from the owner reference in `pod-complete` stage.

https://github.com/NVIDIA/knavigator/blob/main/charts/overrides/kwok/pod-complete.yml

**Reply by yuanchen8911** [reply]
We want to keep a job running and control when to stop it. This can be achieved by setting a very large duration, e.g., 1 year. 

As we discussed internally, there may be a better solution for this use case, but KWOK needs to stay general and flexible  for different use cases.

**Reply by dmitsh** [reply]
Setting a one-year expiration on a pod to allow indefinite operation is neither convenient nor user-friendly.
Additionally, the behavior of 'pod-complete' stage is inconsistent. It only completes the pod if it originates from a job. If the pod originates from another object, such as a replica set, it remains in the running state.
Nevertheless, we have devised a workaround.

**Reply by yuanchen8911** [reply]
To address @dmitsh's concerns regarding "_neither convenient nor user-friendly_", we have three options:

- Options 1: if KWOK stages can function as parametrizable templates, we can customize the default delay in the `pod-complete` stage with an arbitrary value. Thus, we won't need a change (i.e., adding annotations) to a pod. Would this be possible, @wzshiming?

- Option 2: kwok adds an additional stage file `pod-never-complete.yaml` and allows a user to specify which stage(`pod-never-complete.yaml` or `pod-never-complete.yaml`) to deploy? Can helm chart support it? 

- Options 3: a user can always take the current `pod-complete.yaml` and set the default delay to a different value in their internal use. In our case, is this a simpler and safer workaround than removing the `ownerReference`? @dmitsh 

Regardless, the current pod annotation approach works. We just need to add an additional pod annotation to the job templates. 

The pod-complete stage only applies to a pod with the [ownerReferences.kind: job](https://github.com/kubernetes-sigs/kwok/blob/a284d0ca8703257a9b36a48bb4de328898268162/kustomize/stage/pod/fast/pod-complete.yaml#L17). Therefore, a pod from a deployment will remain in the "Running" state until deletion.

**Comment by wzshiming** [comment]
> The delay field in a stage file is specified in milliseconds. I suggest changing it to seconds.
> - Do we really need this level of granularity?

You have the same idea as my original. 😄

The millisecond makes sense to achieve more fine-grained control over resource behavior, although it is not needed in most cases.

Here's the previous discussion
https://github.com/kubernetes-sigs/kwok/pull/74#discussion_r1062975006



This was when Milliseconds was finally chosen for convenience.

I actually think that's probably a better definition, there wasn't enough time.
ref https://github.com/kubernetes/kubernetes/issues/76951#issuecomment-486465014

``` yaml
delayDuration:
  value: <uint64>
  valueFrom:
    expression: <expression-string>
  unit: <unit-string>
```

> - We want to set a very large delay, but the default maximum value in YAML is 999,999. 999,999 milliseconds is too short for our tests.

YAML should have no limit on a number, and even if it did, it would be the maximum of uint64 9223372036854775808 (1<<64-1)

**Reply by yuanchen8911** [reply]
You are right. The maximum value is great enough.  I just tried a value `31536000000` (1 year). It seems to be working. 

 ```
  pod-complete.stage.kwok.x-k8s.io/delay: "31536000000"
   pod-complete.stage.kwok.x-k8s.io/jitter: "31536000000"
```
